### PR TITLE
feat(query-builders): add else support to IFs

### LINKED ISSUE #223: Optional third parameter to .if()
The idea is to run an alternative callback if the condition evaluates to false.

### PR DESCRIPTION
**Pull request by naorpeled**
**TODO**:
- [ ] update docs with examples of elseFunc
- [ ] add tests to `select.test.ts`

closes #223

### CODE REVIEW

File: src/query-builder/delete-query-builder.ts [hunk 1]
```diff
@@ -537,7 +538,8 @@ export class DeleteQueryBuilder<DB, TB extends keyof DB, O>
    */
   if<O2>(
     condition: boolean,
-    func: (qb: this) => DeleteQueryBuilder<DB, TB, O2>
+    func: (qb: this) => DeleteQueryBuilder<DB, TB, O2>,
```
Comment (igalklebanov) [inline src/query-builder/delete-query-builder.ts:541]:
```suggestion
    thenFunc: (qb: this) => DeleteQueryBuilder<DB, TB, O2>,
```

File: src/query-builder/delete-query-builder.ts [hunk 2]
```diff
@@ -537,7 +538,8 @@ export class DeleteQueryBuilder<DB, TB extends keyof DB, O>
...
+    elseFunc?: (qb: this) => DeleteQueryBuilder<DB, TB, O2>
```
Comment (igalklebanov) [inline src/query-builder/delete-query-builder.ts:542]: This won't work, type-wise, if `elseFunc`'s result differs from `func`'s.

File: test/node/src/update.test.ts [hunk 3]
```diff
@@ -209,6 +209,20 @@ for (const dialect of BUILT_IN_DIALECTS) {
...
       })
 
+      it('conditional returning statement that evaluates to false should add optional fields', async () => {
+        const condition = true
+
+        const query = ctx.db
+          .updateTable('person')
+          .set({ last_name: 'Barson' })
+          .returning('first_name')
+          .if(condition, (qb) => qb, (qb) => qb.returning('last_name'))
```
Comment (igalklebanov) [inline test/node/src/update.test.ts:219]: This example doesn't make sense..

File: test/node/src/insert.test.ts [hunk 4]
```diff
@@ -568,6 +568,23 @@ for (const dialect of BUILT_IN_DIALECTS) {
...
+
+          const query = ctx.db
+            .insertInto('person')
+            .values({
+              first_name: 'Foo',
+              last_name: 'Barson',
+              gender: 'other',
+            })
+            .returning('first_name')
+            .if(condition, (qb) => qb, (qb) => qb.returning('last_name'))
```
Comment (igalklebanov) [inline test/node/src/insert.test.ts:583]: This example doesn't make sense..

File: src/query-builder/insert-query-builder.ts [hunk 5]
```diff
@@ -583,7 +584,8 @@ export class InsertQueryBuilder<DB, TB extends keyof DB, O>
    */
   if<O2>(
     condition: boolean,
-    func: (qb: this) => InsertQueryBuilder<DB, TB, O2>
+    func: (qb: this) => InsertQueryBuilder<DB, TB, O2>,
```
Comment (igalklebanov) [inline src/query-builder/insert-query-builder.ts:587]:
```suggestion
    thenFunc: (qb: this) => InsertQueryBuilder<DB, TB, O2>,
```

File: src/query-builder/insert-query-builder.ts [hunk 6]
```diff
@@ -583,7 +584,8 @@ export class InsertQueryBuilder<DB, TB extends keyof DB, O>
...
+    elseFunc?: (qb: this) => InsertQueryBuilder<DB, TB, O2>
```
Comment (igalklebanov) [inline src/query-builder/insert-query-builder.ts:588]: See comments on DeleteQueryBuilder.if..

File: src/query-builder/select-query-builder.ts [hunk 7]
```diff
@@ -1540,12 +1541,17 @@ export class SelectQueryBuilder<DB, TB extends keyof DB, O>
    */
   if<O2 extends O>(
     condition: boolean,
-    func: (qb: this) => SelectQueryBuilder<DB, TB, O2>
-  ): SelectQueryBuilder<DB, TB, MergePartial<O, O2>> {
+    func: (qb: this) => SelectQueryBuilder<DB, TB, O2>,
```
Comment (igalklebanov) [inline src/query-builder/select-query-builder.ts:1544]:
```suggestion
    thenFunc: (qb: this) => SelectQueryBuilder<DB, TB, O2>,
```

File: src/query-builder/select-query-builder.ts [hunk 8]
```diff
@@ -1540,12 +1541,17 @@ export class SelectQueryBuilder<DB, TB extends keyof DB, O>
...
+    elseFunc?: (qb: this) => SelectQueryBuilder<DB, TB, O2>
+    ): SelectQueryBuilder<DB, TB, MergePartial<O, O2>> {
```
Comment (igalklebanov) [inline src/query-builder/select-query-builder.ts:1546]: See comments on DeleteQueryBuilder.if..

File: src/query-builder/update-query-builder.ts [hunk 9]
```diff
@@ -632,7 +633,8 @@ export class UpdateQueryBuilder<DB, UT extends keyof DB, TB extends keyof DB, O>
    */
   if<O2>(
     condition: boolean,
-    func: (qb: this) => UpdateQueryBuilder<DB, UT, TB, O2>
+    then: (qb: this) => UpdateQueryBuilder<DB, UT, TB, O2>,
```
Comment (igalklebanov) [inline src/query-builder/update-query-builder.ts:636]:
```suggestion
    thenFunc: (qb: this) => UpdateQueryBuilder<DB, UT, TB, O2>,
```

File: src/query-builder/update-query-builder.ts [hunk 10]
```diff
@@ -632,7 +633,8 @@ export class UpdateQueryBuilder<DB, UT extends keyof DB, TB extends keyof DB, O>
...
+    elseFunc?: (qb: this) => UpdateQueryBuilder<DB, UT, TB, O2>
```
Comment (igalklebanov) [inline src/query-builder/update-query-builder.ts:637]: See comments on DeleteQueryBuilder.if...

File: src/query-builder/update-query-builder.ts [hunk 11]
```diff
@@ -644,7 +646,11 @@ export class UpdateQueryBuilder<DB, UT extends keyof DB, TB extends keyof DB, O>
       : MergePartial<O, O2>
   > {
     if (condition) {
-      return func(this) as any
+      return then(this) as any
```
Comment (igalklebanov) [inline src/query-builder/update-query-builder.ts:649]:
```suggestion
      return thenFunc(this) as any
```

File: test/node/src/delete.test.ts [hunk 12]
```diff
@@ -161,6 +161,31 @@ for (const dialect of BUILT_IN_DIALECTS) {
...
       })
+
+      it('conditional returning statement that evaluates to false should add optional fields', async () => {
+        const condition = false
+
+        const query = ctx.db
+          .deleteFrom('person')
+          .where('gender', '=', 'female')
+          .returning('first_name')
+          .if(condition, (qb) => qb , (qb) => qb.returning('last_name'))
```
Comment (igalklebanov) [inline test/node/src/delete.test.ts:172]: Weird example..

File: test/node/src/update.test.ts [hunk 13]
```diff
@@ -209,6 +209,20 @@ for (const dialect of BUILT_IN_DIALECTS) {
...
+
+        const result = await query.executeTakeFirstOrThrow()
+        expect(result.last_name).to.equal('Barson')
```
Comment (igalklebanov) [inline test/node/src/update.test.ts:222]: this should be undefined when condition is truthy..

File: src/query-builder/delete-query-builder.ts [hunk 14]
```diff
@@ -551,6 +553,10 @@ export class DeleteQueryBuilder<DB, TB extends keyof DB, O>
       return func(this) as any
```
Comment (igalklebanov) [inline src/query-builder/delete-query-builder.ts:553]:
```suggestion
      return thenFunc(this) as any
```

File: src/query-builder/insert-query-builder.ts [hunk 15]
```diff
@@ -548,6 +548,7 @@ export class InsertQueryBuilder<DB, TB extends keyof DB, O>
 
   /**
    * Call `func(this)` if `condition` is true.
```
Comment (igalklebanov) [inline src/query-builder/insert-query-builder.ts:550]:
```suggestion
   * Call `thenFunc(this)` if `condition` is true.
```

File: src/query-builder/delete-query-builder.ts [hunk 16]
```diff
@@ -502,6 +502,7 @@ export class DeleteQueryBuilder<DB, TB extends keyof DB, O>
 
   /**
    * Call `func(this)` if `condition` is true.
```
Comment (igalklebanov) [inline src/query-builder/delete-query-builder.ts:504]:
```suggestion
   * Call `thenFunc(this)` if `condition` is true.
```

File: src/query-builder/select-query-builder.ts [hunk 17]
```diff
@@ -1490,6 +1490,7 @@ export class SelectQueryBuilder<DB, TB extends keyof DB, O>
 
   /**
    * Call `func(this)` if `condition` is true.
```
Comment (igalklebanov) [inline src/query-builder/select-query-builder.ts:1492]:
```suggestion
   * Call `thenFunc(this)` if `condition` is true.
```

File: src/query-builder/update-query-builder.ts [hunk 18]
```diff
@@ -596,6 +596,7 @@ export class UpdateQueryBuilder<DB, UT extends keyof DB, TB extends keyof DB, O>
 
   /**
    * Call `func(this)` if `condition` is true.
```
Comment (igalklebanov) [inline src/query-builder/update-query-builder.ts:598]:
```suggestion
   * Call `thenFunc(this)` if `condition` is true.
```

**Comment by igalklebanov** [comment]
Thanks! 🍻 

Some concerns:

* could be easily achieved with 2 `.if` invocations or a single `.call` invocation, with plain javascript and assignment.
* could be a pandora's box, people might start asking for "deeper" boolean logic.

**Comment by naorpeled** [comment]
@igalklebanov thanks for the comments, didn't finish the PR yet :P

Great comments though

**Comment by koskimas** [comment]
I don't think it makes sense to implement this feature because of the types. We cannot know which function is executed, so we'd need to always return the union of both branches' output types. Considering you can already do this using another `if` call, the added complexity doesn't make sense.

Sorry for wasting time on this. We need to start tagging issues with something like "ready for implementation" once we've decided it's something that can and should be done.

**Comment by igalklebanov** [comment]
> We need to start tagging issues with something like "ready for implementation" once we've decided it's something that can and should be done.

Added `greenlit` label to the 🧰.

**Comment by naorpeled** [comment]
> I don't think it makes sense to implement this feature because of the types. We cannot know which function is executed, so we'd need to always return the union of both branches' output types. Considering you can already do this using another `if` call, the added complexity doesn't make sense.
> 
> Sorry for wasting time on this. We need to start tagging issues with something like "ready for implementation" once we've decided it's something that can and should be done.

All good 😎

**Pull request closed without merging**